### ISSUE TITLE
[Feat] Cloudwatch support for filling in missing points/gaps

### ISSUE THREAD

**Issue by luispabon** [grafana/grafana#12949]
### What Grafana version are you using?
Hosted grafana.net

### What datasource are you using?
AWS Cloudwatch

### What OS are you running grafana on?
n/a

### What did you do?
I set up a graph with an AWS cloudwatch data source like on the screenshot below, to monitor several backend response types.

When no data is received (eg no 5xx errors on the ALB), the diagram plots a minimum value of 1.

### What was the expected result?
The graph minimum should be zero

### What happened instead?
The graph minimum is one.

It looks like the graph thinks `"no data" === 1`, but when there's data the numbers are good. See screenies below:

![image](https://user-images.githubusercontent.com/6388823/44219109-22165500-a173-11e8-9fa5-46ce90d3d2f6.png)

![image](https://user-images.githubusercontent.com/6388823/44218953-b7651980-a172-11e8-90bc-756ca368b674.png)

Graph's Y minimum is set to nothing, but changing to zero does not help.

I have some additional SingleStat widgets pointed to the exact same metrics and they show 0 when there's no data, which is correct.

**Comment by torkelo** [comment]
Can you use query inspector to view the raw response from cloudwatch? If Grafana shows it like this then it's pretty likely that cloudwatch returns it non zero

**Comment by luispabon** [comment]
@torkelo I just did. For the 5xx plot I get the following:

```json
"response": {
    "results": {
      "A": {
        "refId": "A",
        "series": [
          {
            "name": "HTTP 5xx count",
            "points": [],
            "tags": {
              "LoadBalancer": "xxx/xxx/xxx"
            }
          }
        ],
        "tables": null
      },
```

Yet I get `1.00` on the graph in a place where it should be zero. If I zoom in exactly on that area, then I get nothing at all, which is weird (same stuff on the query inspector). It feels like when it gets no data, it simply uses the last known value, for as long as it's actually on the graph plotted.

If I change the zoom level,

**Comment by marefr** [comment]
@luispabon Please include query inspector result for the graph you included in issue description, i.e. same time range.

**Comment by luispabon** [comment]
```

**Comment by marefr** [comment]
@luispabon thanks for the detailed information. I think your problem is that you've configured Display tab -> Stacking & Null value -> Null value=connected. Please change to null or null as zero. Please let me know if that resolves your problem.

**Comment by luispabon** [comment]
@marefr unfortunately these are already `null` as per defaults, changing to the other two values doesn't solve the problem

**Comment by marefr** [comment]
@luispabon okay. Your problem is basically that only one point is returned from one of the series. #8686 is related. Normally our other core datasources like Graphite, InfluxDB etc have functionality for filling in "missing" points, but problem is that CloudWatch datasource doesn't currently have that functionality. 

One alternative would be to change this issue to a feature request for supporting filling in missing points for CloudWatch. @mtanda do you have any additional thoughts around possibility of filling in missing values using CloudWatch?

**Comment by mtanda** [comment]
It is easy to add the feature. In some case, it is useful, I want the feature too.

**Comment by marefr** [comment]
@luispabon changed title and labeled this as a feature request

**Comment by luispabon** [comment]
@marefr thank you

**Comment by mtanda** [comment]
From ver 5.3, Grafana support `Metric Math` and can use `FILL()` for filling missing points.
https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html

It use `GetMetricData` API. If graph target is only one `Statistics`, it works well.
If user want to show multiple `Statistics`, Grafana need to support filling in backend datasource.

**Comment by marefr** [comment]
@luispabon Please let me know if targeting only one `Statistics` is enough - in that case I'll close this issue.

**Comment by dimrozakis** [comment]
There are several cases that currently make it hard to accomplish using `FILL()`, such as:
1. filling multiple `Statistics` (can't be done according to https://github.com/grafana/grafana/issues/12949#issuecomment-426130177), would probably need each statistic to be defined as distinct query?)
2. filling all/multiple queries (would require multiple maths expressions, one per query)
3. filling queries that use a multi-value template variable as dimension (Though this seems to work in grafana for aggregate queries like `SUM(FILL(METRICS(),0))` it fails for `FILL(METRICS(),0)`. The later does work in CloudWatch Console).

If `FILL(METRICS(),0)` worked properly, it'd solve points 2 & 3 from above, leaving only the case for multiple stats.

**Comment by MartinLoeper** [comment]
I have a similar issue: Two sparse metrics (LoginCount_Sum with id:a and LoginFailureCount_Sum with id:b) which should be filled with 0 if not available. Unfortunately, as already stated, CloudWatch does not return null or 0 values for custom sparse metrics.

I solved it by using a combination of math expressions (one for each query) and series overrides (one for each query).

<img width="868" alt="metric" src="https://user-images.githubusercontent.com/5209395/58637768-8d40aa80-82f3-11e9-9593-7e30c7152ae7.PNG">
<img width="863" alt="visualization" src="https://user-images.githubusercontent.com/5209395/58637781-93368b80-82f3-11e9-8e23-69a2186fdda9.PNG">

The stacked bar chart renders seamlessly now:
<img width="942" alt="chart" src="https://user-images.githubusercontent.com/5209395/58638088-41dacc00-82f4-11e9-9002-f880f50551ee.PNG">


### Why do I need this?
If I do not fill the metric with 0, the bar chart renders strangely in some circumstances (the bars get very wide). I think this is because the value stays the same between two data points.

**Comment by chipironcin** [comment]
+1 to this
I can't try the "SUM,FILL,METRICS" hack because I've got multiple metrics aggregated (dynamic metric definition).

The values Grafana shows might be misleading as current values for some metrics are mixed with older values for other metrics, not showing the real count.

**Comment by sunker** [comment]
A lot of changes have been made the cloudwatch datasource in grafana 6.5 (current released version is 6.5-beta1). One of those changes was to move away from the GetMetricStatistics API and use GetMetricData for all requests. See #20057 From now on, at any time you use a multi-value template variable for dimension value, or you use a wildcard `*` for a dimension, a search expression is being generated by grafana. You can preview that search expression by clicking `Show query preview`. The generated search expression is using a new cloudwatch function `REMOVE_EMPTY` which will throw away all time series that are empty. If this won't be enough, you can just copy the search expression from the query preview and put it in the expression field. Then add appropriate functions. e.g `REMOVE_EMPTY(SUM(FILL(METRICS(),0)))` Read more about search expressions [here](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-search-expressions.html)
Closing this issue now
